presto-main/src/main/java/com/facebook/presto/operator/BigintInMemoryJoinHash.java

```diff
@@ -0,0 +1,232 @@
+                    break;
+                }
+                // increment position and mask to handler wrap around
+                hashPos = (hashPos + 1) & mask;
+                //currentKey = key.get(hashPos);
+                currentKey = key[hashPos];
+            }
+
+            key[hashPos] = position;
+        }
+
+        totalMemoryUsage = sizeOf(key) + sizeOf(values)
+                +  sizeOf(addresses.elements()) + pagesHashStrategy.getSizeInBytes()
+                + (positionLinks.isPresent() ? sizeOf(positionLinks.get()) : 0)
+                + (nullsMask.isPresent() ? sizeOf(nullsMask.get().toLongArray()) : 0);
+    }
+
+    private void setPositionLinks(int position, int currentKey)
+    {
+        if (!positionLinks.isPresent()) {
+            positionLinks = Optional.of(new int[addresses.size()]);
+            Arrays.fill(positionLinks.get(), -1);
+        }
+
+        positionLinks.get()[position] = currentKey;
+    }
+
+    private void extractValues(LongArrayList addresses, List<Block> joinBlocks)
+    {
+        for (int position = 0; position < addresses.size(); position++) {
+            long pageAddress = addresses.getLong(Ints.checkedCast(position));
+            int blockIndex = decodeSliceIndex(pageAddress);
+            int blockPosition = decodePosition(pageAddress);
+
+            long value;
+            if (joinBlocks.get(blockIndex).isNull(blockPosition)) {
+                if (!nullsMask.isPresent()) {
+                    nullsMask = Optional.of(new BitSet(addresses.size()));
+                }
+                nullsMask.get().set(position, true);
+                value = Long.MIN_VALUE;
+            }
+            else {
+                value = BIGINT.getLong(joinBlocks.get(blockIndex), blockPosition);
+            }
+            values[position] = value;
+        }
+    }
+
+    @Override
+    public final int getChannelCount()
+    {
+        return channelCount;
+    }
+
+    @Override
+    public int getJoinPositionCount()
+    {
+        return addresses.size();
+    }
+
+    @Override
+    public long getInMemorySizeInBytes()
+    {
+        return totalMemoryUsage;
+    }
+
+    @Override
+    public long getJoinPosition(int position, Page page, int rawHash)
+    {
+        return getJoinPosition(position, page);
+    }
+
+    @Override
+    public long getJoinPosition(int position, Page page)
+    {
+        long probeValue;
+        boolean isNull;
+        if (page.getBlock(0).isNull(position)) {
+            probeValue = Long.MIN_VALUE;
+            isNull = true;
+        }
+        else {
+            probeValue = BIGINT.getLong(page.getBlock(0), position);
+            isNull = false;
+        }
+        int hashPosition = getHashPosition(probeValue, mask);
+        int buildPosition = key[hashPosition];
+        while (buildPosition != -1) {
+            long buildValue = values[buildPosition];
+
+            if (buildValue == probeValue && bothNullOrNonNull(buildPosition, isNull)) {
```

Comment:
That's the semantic of `com.facebook.presto.type.TypeUtils#positionEqualsPosition` that was previously being used for BIGINT joins (and all the other joins)
```
    public static boolean positionEqualsPosition(Type type, Block leftBlock, int leftPosition, Block rightBlock, int rightPosition)
    {
        boolean leftIsNull = leftBlock.isNull(leftPosition);
        boolean rightIsNull = rightBlock.isNull(rightPosition);
        if (leftIsNull || rightIsNull) {
            return leftIsNull && rightIsNull;
        }
        return type.equalTo(leftBlock, leftPosition, rightBlock, rightPosition);
    }
```